### PR TITLE
init v2

### PR DESCRIPTION
**Pull request by webblin034**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+'use strict';
+
+var seoInspector = require('./build/SeoInspector.js');
```
Comment (mumu-ilu) [inline index.js]: Class must be "PascalCased"

File: gulpfile.babel.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import gulp from 'gulp';
+import babel from 'gulp-babel';
+
+gulp.task('Watch', () => {
+    gulp.watch(['src/*.js', 'src/rules/*.js', 'tests/*.js'], ['TranspileEs6ToEs5']);
+});
+
+gulp.task('TranspileEs6ToEs5', () => {
+    gulp.src('src/*.js')
+        .pipe(babel({ presets: ['es2015'], plugins: ['add-module-exports'] }))
```
Comment (mumu-ilu) [inline gulpfile.babel.js:12]: Is this configuration necessary ? .babelrc will handle it ~

File: src/SeoInspector.js [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            dom => this._getSummary(dom)
+        )
+        .then(
+            summary => this._output(summary)
+        );
+    }
+
+    _getHtml() {
+    	return new Promise((resolve, reject) => {
+    	    const type = this.input.constructor.name;
```
Comment (mumu-ilu) [inline src/SeoInspector.js]: "Duck type checking" is the best way to go.
`if (typeof type === 'string') // for case 'String'`
`if (type && typeof type.on === 'function') // for case 'ReadStream'`

File: src/SeoInspector.js [hunk 4]
```diff
@@ -0,0 +1,148 @@
+'use strict';
+
+import fs from 'fs';
+import { JSDOM } from 'jsdom';
+import _ from 'lodash';
+import conv from 'iconv-lite';
+import RuleFactory from './RuleFactory';
+
+const defaultRules = {
+    'ImgTagWithAltAttritube': { enabled: 1 },
```
Comment (mumu-ilu) [inline src/SeoInspector.js]: To avoid unexpected error (e.g., typo), we should use string constant here.

File: src/SeoInspector.js [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    	    const type = this.input.constructor.name;
+    	    switch (type) {
+    	        case 'String':
+    	            const filePath = this.input;
+    	            fs.readFile(filePath, 'utf8', (err, html) => {
+    	                if (err) return this.done(err);
+    	                resolve(html);
+    	            });
+    	            break;
+    	        case 'ReadStream':
```
Comment (mumu-ilu) [inline src/SeoInspector.js]: What if exception occurs during stream reading ?

File: src/SeoInspector.js [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+class SeoInspector {
+
+    constructor(options = {}) {
+        this.done = (options.done) ? options.done : (err) => { if (err) throw err };
+        this.input = null;
+        this.output = null;
+        this.rules = {};
+        this.ruleFactory = new RuleFactory();
+        return this;
```
Comment (mumu-ilu) [inline src/SeoInspector.js]: Why return 'this' ref here?

File: src/SeoInspector.js [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    read(input) {
+        this.input = input;
+        return this;
```
Comment (mumu-ilu) [inline src/SeoInspector.js]: Same as above.

File: src/SeoInspector.js [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        });
+    }
+    
+    _getSummary(dom) {
+        return new Promise((resolve, reject) => {
+            let summary = [];
+            _.forEach(this.rules, (options, name) => {
+                let ruleObj = this.ruleFactory.create(name, options, this.done);
+                let report = ruleObj.check(dom)
+                if (report && report != 'undefined') {
```
Comment (mumu-ilu) [inline src/SeoInspector.js]: null, undefined and empty string are all 'falsy value'

According to your ruleObject implementation, you can check by this:

`if (!report)`

File: src/rules/ATagWithRelAttritubeRule.js [hunk 9]
```diff
@@ -0,0 +1,27 @@
+'use strict';
+
+import BaseRule from './BaseRule';
+
+class ATagWithRelAttritubeRule extends BaseRule {
+
+    check(dom) {
+        if (this.options.enabled == 0) {
```
Comment (mumu-ilu) [inline src/rules/ATagWithRelAttritubeRule.js]: use `===` instead of `==`

File: src/rules/HeadTagWithTitleAndDescriptionsKeywordsMetaRule.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
+'use strict';
+
+import BaseRule from './BaseRule';
+
+class HeadTagWithTitleAndDescriptionsKeywordsMetaRule extends BaseRule {
+
+    check(dom) {
+        if (this.options.enabled == 0) {
```
Comment (mumu-ilu) [inline src/rules/HeadTagWithTitleAndDescriptionsKeywordsMetaRule.js]: use `===` instead of `==`

File: src/rules/ImgTagWithAltAttritubeRule.js [hunk 11]
```diff
@@ -0,0 +1,27 @@
+'use strict';
+
+import BaseRule from './BaseRule';
+
+class ImgTagWithAltAttritubeRule extends BaseRule {
+
+    check(dom) {
+        if (this.options.enabled == 0) {
```
Comment (mumu-ilu) [inline src/rules/ImgTagWithAltAttritubeRule.js]: use `===` instead of `==`

File: src/rules/NoMoreThanOneH1TagRule.js [hunk 12]
```diff
@@ -0,0 +1,21 @@
+'use strict';
+
+import BaseRule from './BaseRule';
+
+class NoMoreThanOneH1Tag extends BaseRule {
+
+    check(dom) {
+        if (this.options.enabled == 0) {
```
Comment (mumu-ilu) [inline src/rules/NoMoreThanOneH1TagRule.js]: use `===` instead of `==`

File: src/rules/NoTooManyStrongTagsRule.js [hunk 13]
```diff
@@ -0,0 +1,21 @@
+'use strict';
+
+import BaseRule from './BaseRule';
+
+class NoTooManyStrongTagsRule extends BaseRule {
+
+    check(dom) {
+        if (this.options.enabled == 0) {
```
Comment (mumu-ilu) [inline src/rules/NoTooManyStrongTagsRule.js]: use `===` instead of `==`

File: src/rules/ATagWithRelAttritubeRule.js [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+            return;
+        }
+        let report = '';
+        let count = 0;
+        const elements = dom.window.document.querySelectorAll("a");
```
Comment (mumu-ilu) [inline src/rules/ATagWithRelAttritubeRule.js]: use single quote here.

File: src/SeoInspector.js [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+                    summary.push(report);
+                }
+            });
+            resolve(summary);
+        });
+    };
+    
+    _output(summary) {
+        return new Promise((resolve, reject) => {
+            let contentStr = "";
```
Comment (mumu-ilu) [inline src/SeoInspector.js]: use single quote here.

File: src/rules/ImgTagWithAltAttritubeRule.js [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+            return;
+        }
+	let report = '';
```
Comment (mumu-ilu) [inline src/rules/ImgTagWithAltAttritubeRule.js]: code indentation

**Review by mumu-ilu — CHANGES_REQUESTED** [review]

File: src/SeoInspector.js [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+    'ATagWithRelAttritube': { enabled: 1 },
+    'HeadTagWithTitleAndDescriptionsKeywordsMeta': { enabled: 1 },
+    'NoTooManyStrongTags': { enabled: 1, threshold: 15 },
+    'NoMoreThanOneH1Tag': { enabled: 1 }
+};
+
+class SeoInspector {
+
+    constructor(options = {}) {
+        this.done = (options.done) ? options.done : (err) => { if (err) throw err };
```
Comment (mumu-ilu) [inline src/SeoInspector.js]: type checking is good enough here ?

File: src/SeoInspector.js [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+    
+    _output(summary) {
+        return new Promise((resolve, reject) => {
+            let contentStr = "";
+            if (summary && summary.length > 0) {
+                contentStr = "SEO defects found: \n" + summary.join("\n") + "\n";
+            } else {
+                contentStr = "No any SEO defect found. \n"; 
+            }
+            const type = this.output.constructor.name;
```
Comment (mumu-ilu) [inline src/SeoInspector.js]: use duck type checking here

File: src/rules/HeadTagWithTitleAndDescriptionsKeywordsMetaRule.js (hunk 10 above)
Comment (mumu-ilu) [inline src/rules/HeadTagWithTitleAndDescriptionsKeywordsMetaRule.js]: This logic should be 'inherited' from parent class 'BaseRule' in order to reduce redundant code. Let's use method override
`// Parent BaseRule`
`check(dom) { if (this.options.enable === 0) return }`

`// Child xxxxRule`
`check(dom) { super.check(dom) }`

File: index.js (hunk 1 above)
Comment (webblin034) [reply index.js]: done

File: src/SeoInspector.js (hunk 5 above)
Comment (webblin034) [reply src/SeoInspector.js]: readStream.on('error', (e) => {
                        this.done(e);
                    });

File: src/SeoInspector.js (hunk 6 above)
Comment (webblin034) [reply src/SeoInspector.js]: for chaining
new Inspector().read().addRule().write();

File: src/SeoInspector.js (hunk 7 above)
Comment (webblin034) [reply src/SeoInspector.js]: for chainning as above

File: src/SeoInspector.js (hunk 8 above)
Comment (webblin034) [reply src/SeoInspector.js]: done => if (report) summary.push

File: src/rules/HeadTagWithTitleAndDescriptionsKeywordsMetaRule.js (hunk 10 above)
Comment (webblin034) [reply src/rules/HeadTagWithTitleAndDescriptionsKeywordsMetaRule.js]: done

File: src/rules/ImgTagWithAltAttritubeRule.js (hunk 11 above)
Comment (webblin034) [reply src/rules/ImgTagWithAltAttritubeRule.js]: done

File: src/rules/NoMoreThanOneH1TagRule.js (hunk 12 above)
Comment (webblin034) [reply src/rules/NoMoreThanOneH1TagRule.js]: done

File: src/rules/NoTooManyStrongTagsRule.js (hunk 13 above)
Comment (webblin034) [reply src/rules/NoTooManyStrongTagsRule.js]: done

File: src/SeoInspector.js (hunk 15 above)
Comment (webblin034) [reply src/SeoInspector.js]: done

File: src/rules/ATagWithRelAttritubeRule.js (hunk 14 above)
Comment (webblin034) [reply src/rules/ATagWithRelAttritubeRule.js]: done

File: src/rules/ImgTagWithAltAttritubeRule.js (hunk 16 above)
Comment (webblin034) [reply src/rules/ImgTagWithAltAttritubeRule.js]: ok

File: src/SeoInspector.js (hunk 17 above)
Comment (webblin034) [reply src/SeoInspector.js]: this.done = (_.has(options, ['done'])) ? options.done : (err) => { if (err) throw err };

File: src/rules/HeadTagWithTitleAndDescriptionsKeywordsMetaRule.js (hunk 10 above)
Comment (webblin034) [reply src/rules/HeadTagWithTitleAndDescriptionsKeywordsMetaRule.js]: return 的行為
好像無法用繼承來實作減少 duplicated
sub class 即使 呼叫 super.check 也無法中斷

**Pull request merged**